sematic/graph.py

```diff
@@ -0,0 +1,560 @@
+
+            future = func(**kwargs)
+
+            input_artifacts[future.id] = run_input_artifacts
+
+            if run_id not in reset_run_ids:
+                value, output_artifact = _get_run_output(run_id)
+                if output_artifact is not None:
+                    output_artifacts[future.id] = output_artifact
+                    future.value = value
+
+            _set_parent_nested(future, run)
+
+            # Settings state for resolved runs unless reset
+            if FutureState[run.future_state] == FutureState.RESOLVED:  # type: ignore
+                if run.id not in reset_run_ids:
+                    future.state = FutureState.RESOLVED
+
+            futures_by_original_id[run.id] = future
+
+        for run_id in run_ids_by_execution_order:
+            if run_id in skip_run_ids:
+                continue
+
+            _clone_run(run_id)
+
+        if reset_from is None and len(futures_by_original_id) != len(self.runs):
+            raise RuntimeError("Not all futures duplicated")
+
+        # We return future sorted by how they would be sorted for a resolution
+        # from scratch: grouped by layer (outermost first), and sorter in reverse
+        # execution order within each layer
+        run_ids_by_reverse_execution_order = self._sorted_run_ids_by_layer(
+            run_sorter=self._reverse_execution_order
+        )
+
+        return (
+            collections.OrderedDict(
+                (
+                    (run_id, futures_by_original_id[run_id])
+                    for run_id in run_ids_by_reverse_execution_order
+                    if run_id in futures_by_original_id
+                )
+            ),
+            input_artifacts,
+            output_artifacts,
+        )
+
+
+EdgeFilterCallable = Callable[[List[Optional[RunID]], RunID], bool]
+
+
+def _upstream_edge_filter(
+    edges_by_destination_id: Dict[RunID, List[Edge]]
+) -> EdgeFilterCallable:
+    def _edge_filter(upstream_run_ids: List[Optional[RunID]], run_id: RunID):
+        return all(
+            edge.source_run_id in upstream_run_ids
+            for edge in edges_by_destination_id[run_id]
+        )
+
+    return _edge_filter
+
+
+def _downstream_edge_filter(
+    edges_by_source_id: Dict[RunID, List[Edge]]
+) -> EdgeFilterCallable:
+    def _edge_filter(downstream_run_ids: List[Optional[RunID]], run_id: RunID):
+        return all(
+            edge.destination_run_id in downstream_run_ids
+            for edge in edges_by_source_id[run_id]
+        )
+
+    return _edge_filter
+
+
+def _sort_layer_runs(
+    layer_run_ids: List[RunID], edge_filter: EdgeFilterCallable
+) -> List[str]:
+    """
+    Sort runs topologically within layer. The order (upstream first or
+    downstream first) depends on edge_filter.
+
+    Parameters
+    ----------
+    layer_run_ids: List[RunID]
+        All run IDs in the layer
+    edge_filter: EdgeFilterCallable
+        A callable to determine the next set of run IDs based on their edges.
```

Comment:
reading this, I would not understand what it takes to define an `edge_filter`. It says it's used to determine the next set of run ids, but it accepts a list of run ids, a single run id, and returns a bool. What am I supposed to do with the run ids? If I'm supposed to "determine the next set of run ids" with it, how do I do that by just returning a bool? Shouldn't I be returning a list of run ids? If it was very easy to understand what the expectations were by reading the code, it wouldn't require explicit explanation. But it took me several minutes to read the code and grok it well enough to understand what was going on the first time. and still on the second read I have to digest it and think through it again to recall. Another two or three sentences will save people a lot of time I think.